File: 0496-next-greater-element-i/0496-next-greater-element-i.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        HashMap<Integer, Integer> nge = new HashMap<>();
        Stack<Integer> stk = new Stack<>();
        int n = nums2.length;
        for (int i = n - 1; i >= 0; i--) {
            while (!stk.isEmpty() && stk.peek() <= nums2[i]) {
                stk.pop();
            }
            if (i < n) {
                if (!stk.isEmpty()) {
                    nge.put(nums2[i], stk.peek());
                } else {
                    nge.put(nums2[i], -1);
                }
            }
            stk.push(nums2[i]);
        }
        n = nums1.length;
        int[] res = new int[n];
        for (int i = 0; i < n; i++) {
            res[i] = nge.get(nums1[i]);
        }
        return res;
    }
}

File: 0496-next-greater-element-i/0496-next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        stack<int> st;
        unordered_map<int, int> mp;
        
        for (int i = nums2.size() - 1; i >= 0; i--) {
            while (!st.empty() && nums2[st.top()] <= nums2[i]) {
                st.pop();
            }
            mp[nums2[i]] = st.empty() ? -1 : nums2[st.top()];
            st.push(i);
        }
        
        vector<int> res(nums1.size());
        
        for (int i = 0, n = nums1.size(); i < n; i++) {
            res[i] = mp[nums1[i]];
        }
        
        return res;
    }
};